File: leetcodes/53. Maximum Subarray/maximumSubarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
 var maxSubArray = function(nums) {
   //input: array
   //output: integer (largest sum)
   //edge case: nums.length >=1 return nums[0]
   //test case: [-3,-2,-1] -1, [-2,1,-3,4,-1,2,1,-5,4] 6
  //time: O(n), space O(1)
  //don't need to use negative sum to keep adding up
  if (nums.length === 1) return nums[0];
  let sum = nums[0];
  let max = nums[0];
  for (let i = 1; i<nums.length; i++) {
    //if sum < 0, use next elem as sum
    if (sum < 0) {
      sum = nums[i];
      //if sum > 0, sum adds next elem into sum
    } else {
      sum += nums[i];
    }
    //compare max and sum and replace max as the largest one
    max = Math.max(max, sum);
  }
  return max;
 };

File: leetcodes/53. Maximum Subarray/maximumSubarray.java
class Solution {
  public int maxSubArray(int[] nums) {
    //edge case nums length = 1 return nums[0]
    //iterate over the array, with max and curSum
    if (nums.length == 1) return nums[0];
    int maxAns = nums[0];
    int curSum = nums[0];
    for (int i = 1; i< nums.length; i++) {
      if (curSum < 0) {
        curSum = nums[i];
        //create new start
      } else {
        curSum += nums[i];
      }
      maxAns = Math.max(maxAns, curSum);
    }
    return maxAns;
  }
}